### DISCUSSION TITLE
General Discussion: Multiple Topics

### DISCUSSION THREAD

**Discussion by keerthankrishna** [react-querybuilder/react-querybuilder#446] [General]
Please forgive me if the question is silly - Can i render different value dropdowns based on the field selection ?
I tried doing this using the ` getValues ` function but I think I'm making some mistake. 

_For some reason that function is not getting called at all_
CodeSandbox Link [here](https://codesandbox.io/s/laughing-tom-wlkrup-react-qb-forked-g3lqf7?file=/src/App.js)

My end-goal is the have a GET API that gives me the list of values based on the field selection but in the above Sandbox link, I have just used a sample for testing purpose.

Secondary Question: Can I not use the same approach for a "multi-select" input as well ? I see in the docs, it is mentioned that only "select" and "radio" inputs are compatible with ` getValues `

**Comment by keerthankrishna** [comment]
I think I answered my first question - It works now in my CodeSandbox [link](https://codesandbox.io/s/laughing-tom-wlkrup-react-qb-forked-g3lqf7?file=/src/App.js). I made a mistake in calling the function ` getValues ` but please let me know about the second question.

Also, can I change the _type of input field based on the operator selection_ ? 
**Example**: If i select _in_ operator, I want a dropdown input field to show up and if i select _begins with_ operator, then i want a text field to show up

**Comment by jakeboone02** [comment]
See if this answers your second question: https://codesandbox.io/s/rqb-446-9i0vqg?file=/src/App.js

(The docs have been fixed: see 639d1b513116d1750430fd988cffc62c662d35bf.)

**Reply by keerthankrishna** [reply]
That works! Thank you :)
Is it possible to _change the input field based on the operator selection_ ? (mentioned in my previous reply but here is it incase you missed it as I just edited the response and added it now)

If i select in operator, I want a dropdown input field to show up and if i select begins with operator, then i want a text field to show up

**Reply by jakeboone02** [reply]
I think you'd need a custom `valueEditor` for that since the default value editor doesn't do anything different for the "in" operator. See [tips here](https://react-querybuilder.js.org/docs/tips/custom-with-fallback).

**Reply by keerthankrishna** [reply]
Super! I was able to work it out here using the custom ` valueEditor ` component and it's awesome that you enable so much data to be passed via props to it! I was able to check the field value as well as the operator used before rendering the ` value-editor ` dropdown

**One more question** - Is it possible for me to fetch the selected field anyway and use it somewhere else ? 
_Here's the reason for the question: I want to call an API and pass the selected field input to the API request to fetch a list of values to be shown as dropdown for that selected field._

**Reply by jakeboone02** [reply]
I would suggest using the [`Async` component from `react-select`](https://react-select.com/async). The `loadOptions` callback is where you would pass the `field` to the API.

**Reply by keerthankrishna** [reply]
Can the custom ` valueEditor ` take other props with it ? 
ie; I would like to pass a state value to it when i call it inside ` controlElements ` within ` QueryBuilder `, so that I can use it in the component definition where ` valueEditor ` is defined

**Reply by jakeboone02** [reply]
Best practice would be to use the [`context` prop](https://react-querybuilder.js.org/docs/api/querybuilder#context), which is passed down to all subcomponents including custom ones.

**Reply by keerthankrishna** [reply]
> I think you'd need a custom `valueEditor` for that since the default value editor doesn't do anything different for the "in" operator. See [tips here](https://react-querybuilder.js.org/docs/tips/custom-with-fallback).

Thanks for this - I got it working taking reference from the [tips](https://react-querybuilder.js.org/docs/tips/custom-with-fallback) you shared but I seem to have a problem: 
When I call the ` react-select ` in the IF loop (_basically finding the props that that select mentioned_), the rendering part looks all good and the input field gets converted to a ` react-select ` dropdown as well showing the data I need in the dropdown but it does not seem to communicate back to the ` <QueryBuilder> ` component. The SQL query does not seem to get the value I select here.

Let me see if I can share a _CodeSandbox_ link mimicking the issue

**Reply by keerthankrishna** [reply]
@jakeboone02 Here's the CodeSandBox [link](https://codesandbox.io/s/laughing-tom-wlkrup-react-qb-forked-addedcustomvalue-editor-g3lqf7?file=/src/App.js)

I'm able to replicate it here in CodeSandbox as well.
It may possibly be a defect or I'm doing something wrong. Please let me know.

Whatever I select in `react-select` dropdown, does not show up in the `SQL export`
![image](https://user-images.githubusercontent.com/24373382/211659678-3ebd3528-ea1c-4b49-8056-2fc78cba0d06.png)

**Reply by jakeboone02** [reply]
It just needed a few additional props (see https://codesandbox.io/s/75z8iz?file=/src/App.js):

One in `App.js`:

```diff
 const MyFieldSelector = props => (
   <Select
     options={fields}
     isClearable={true}
     isSearchable={true}
     placeholder="Type to Search"
+    getOptionValue={op => op.name}
     value={fields.find(f => f.name === props.value)}
     onChange={v => props.handleOnChange(v?.name)}
   />
 );
```

And two in `CustomValueEditor.js`:

```diff
 export const CustomValueEditor = ValueEditorProps => {
   console.log(ValueEditorProps);
   if (
     ValueEditorProps.field === 'designation' &&
     ValueEditorProps.fieldData.valueEditorType === 'select' &&
     ValueEditorProps.operator === 'in'
   ) {
     console.log('Designation');
     return (
       <div>
         <Select
           options={designationList}
           isClearable={true}
           isSearchable={true}
+          value={designationList.find(d => d.value === ValueEditorProps.value)}
+          onChange={v => ValueEditorProps.handleOnChange(v.value)}
         />
       </div>
     );
   }
 
   return <ValueEditor {...ValueEditorProps} />;
 };
```

**Reply by keerthankrishna** [reply]
Ah, okay - Thank you so much for correcting it.
I would not have found this out in my own.

**Reply by keerthankrishna** [reply]
What change would be needed on the ` QueryBuilder ` side for it to be compatible with a "multi selection" enabled `react-select` component ? 

I have used the same forked version and set the ` react-select ` to multi-selection here - [Link](https://codesandbox.io/s/laughing-tom-wlkrup-react-qb-forked-addedcustomvalue-editor-7u2811?file=/src/App.js)
![image](https://user-images.githubusercontent.com/24373382/211676358-3af90e92-96c8-4a76-a8fd-83ae9e53d991.png)

![image](https://user-images.githubusercontent.com/24373382/211676571-6eeb57ba-770c-42a6-b621-1edfa06cc208.png)

**Reply by jakeboone02** [reply]
https://codesandbox.io/s/s3rjvc?file=/src/CustomValueEditor.js

```diff
 export const CustomValueEditor = ValueEditorProps => {
   console.log(ValueEditorProps);
   if (
     ValueEditorProps.field === 'designation' &&
     ValueEditorProps.operator === 'in'
   ) {
     console.log('Designation');
     return (
       <div>
         <Select
           options={designationList}
           isClearable={true}
           isMulti={true}
           isSearchable={true}
-          value={designationList.find(d => d.value === ValueEditorProps.value)}
-          onChange={v => ValueEditorProps.handleOnChange(v.value)}
+          value={designationList.filter(d =>
+            ValueEditorProps.value.split(',').includes(d.value)
+          )}
+          onChange={v =>
+            ValueEditorProps.handleOnChange(v.map(val => val.value).join(','))
+          }
         />
       </div>
     );
   }
 
   return <ValueEditor {...ValueEditorProps} />;
 };
```

**Reply by keerthankrishna** [reply]
Thank you! You're awesome!

**Comment by keerthankrishna** [comment]
Can I customize the SQL output based on selection of specific fields/operators ?
**Example:** _Out of a set of fields A,B,C,D.. if I specifically select C and then if I select a custom operator called "count", the SQL output would need to be LEN(C) which is basically the count of C. Is this possible and if yes, what would be the approach I would need to take ? Any directions you can give me ?_

Please let me know if my question is not clear - I will try to explain it further.

**Reply by jakeboone02** [reply]
Have you looked at the [documentation for the `valueProcessor` option](https://react-querybuilder.js.org/docs/api/export#value-processor)?

**Reply by keerthankrishna** [reply]
Great! This partially gives me what I want but it does not seem to overwrite the whole thing.
See this CodeSandBox [link](https://codesandbox.io/s/custom-sql-query-rewrite-n3wpcj?file=/src/App.js:1442-1462).

Here, I'm expecting it to give me **LEN(address) < 45** but the _field_ stays there. Is there a way I can completely strip off the `SQL export ` output and show what i provide in the ` valueProcessor ` return - instead it returns **address < LEN(address) < 45**
![image](https://user-images.githubusercontent.com/24373382/211925762-16ce644d-6f3b-4bbd-8095-6ec2af56f33b.png)

**Reply by jakeboone02** [reply]
Try [`ruleProcessor`](https://react-querybuilder.js.org/docs/api/export#rule-processor) instead.

**Reply by keerthankrishna** [reply]
Is there a SQL version of ` defaultRuleProcessorJsonLogic ` ?

**Reply by jakeboone02** [reply]
Ack...sorry! No, there isn't. I think the reason is that the SQL export actually comes in `"parameterized"` and `"parameterized_named"` flavors, so `formatQuery` doesn't really allow custom output of the entire rule, just the value portion.

**Reply by keerthankrishna** [reply]
ah, okay! No problem.

**Comment by keerthankrishna** [comment]
I have a page which I use for both loading existing queries as well as allow creation of new queries and I manage this with a flag that I pass to the page which tells me if the page is loading an existing query _(where it needs to preload some rules from the database)_ of if it's a page to create a new query _(where it needs to be blank on the rules/groups side)_. 

@jakeboone02 Can you give me some direction on how I can approach this ? I tried to set state with a default `{rules:  []}` and then (based on the flag which I set in the later part of the code) did a "_setState_" where I pull the rules and pass it to this. Problem is, by this time the ` <QueryBuilder> ` component is already loaded, so it's not picking the updated rules it needs to preload, i think.

**Reply by jakeboone02** [reply]
I'd need to see a reproduction to know where the timing is going wrong.

**Reply by keerthankrishna** [reply]
@jakeboone02 Here's an example where I initially set ` initialQuery ` and then re-assign it with a rule but it takes what was first initialized.
https://codesandbox.io/s/pedantic-thunder-b6siqu?file=/src/App.js

**Reply by jakeboone02** [reply]
I'm not really sure what you're trying to do, but here are my comments on your code:

```tsx
export const App = () => {
  // If this is a constant, define it outside the component function.
  // Otherwise it gets redefined on every render.
  let initialQuery = { rules: [] };

  // This next line is the only time `initialQuery` is used,
  // and only on the very first render.
  const [query, setQuery] = useState(initialQuery);

  // This next line is essentially ignored. `query` and `setQuery`
  // are used in the returned JSX below. `query` is _initialized_
  // with `initialQuery`, but can only be updated using `setQuery`.
  initialQuery = {
    rules: [
      {
        field: "firstName",
        operator: "=",
        value: "Steve",
      },
    ],
  };

  return (
    <div>
      <QueryBuilder
        fields={fields}
        query={query}
        onQueryChange={(q) => setQuery(q)}
      />
      <h4>Query</h4>
      <pre>
        <code>{formatQuery(query, "json")}</code>
      </pre>
    </div>
  );
};
```

Based on what you've shown me, this seems more like what you want. The button action could be invoked by anything, e.g. an API call response.

```tsx
const initialQuery = { rules: [] };

export const App = () => {
  const [query, setQuery] = useState(initialQuery);

  const updateQuery = () => {
    setQuery({
      rules: [
        {
          field: "firstName",
          operator: "=",
          value: "Steve",
        },
      ],
    });
  };

  return (
    <div>
      <QueryBuilder
        fields={fields}
        query={query}
        onQueryChange={(q) => setQuery(q)}
      />
      <button type="button" onClick={updateQuery}>
        Update query
      </button>
      <h4>Query</h4>
      <pre>
        <code>{formatQuery(query, "json")}</code>
      </pre>
    </div>
  );
};
```

**Reply by keerthankrishna** [reply]
Thank you so much for the feedback - this worked!

**Comment by keerthankrishna** [comment]
@jakeboone02 Why do I not see the "_combinators_" in the **Output JSON** like how it shows up in your demo ?

Please see the CodeSandBox [link](https://codesandbox.io/s/pedantic-thunder-b6siqu?file=/src/App.js)
In your demo [here](https://react-querybuilder.js.org/demo#addRuleToNewGroups=false&autoSelectField=true&autoSelectOperator=true&debugMode=false&disabled=false&enableDragAndDrop=false&independentCombinators=false&justifiedLayout=false&listsAsArrays=false&parseNumbers=false&resetOnFieldChange=true&resetOnOperatorChange=false&showBranches=false&showCloneButtons=false&showCombinatorsBetweenRules=true&showLockButtons=false&showNotToggle=false&validateQuery=false), you get it like how it's highlighted below:
![image](https://user-images.githubusercontent.com/24373382/215228795-3fa13010-63f1-492d-ac3b-c74c7d84a139.png)

In the CodeSandbox link, I don't seem to be getting the "combinators" header before the combinator value I'm selecting.
![image](https://user-images.githubusercontent.com/24373382/215228876-2f56eb14-cc06-4066-b309-2abdbbcf80ac.png)

**Reply by jakeboone02** [reply]
Since your `initialQuery` doesn't have the `combinators` property, it throws the `QueryBuilder` into "independent combinators" mode (even though you don't have the [`independentCombinators` prop](https://react-querybuilder.js.org/docs/api/querybuilder#independentcombinators) set). FWIW, TypeScript would have caught that problem at compile-time.

The fix is to add `combinator` to your `initialQuery`:

```diff
-  let initialQuery = { rules: [] };
+  let initialQuery = { combinator: 'or', rules: [] };
```

Also you don't have to define `initialQuery` inside the component function since it never changes (speaking of that...you can also use `const` instead of `let` if you want).

**Reply by keerthankrishna** [reply]
Thanks for that! Yes, React did not catch that..

So, I made the changes.... but it seems to be showing the "combinators" only when inside a group but not when it's between 2 rules outside a group. Thoughts ? Am I making any mistake here ?
See the same codeSandbox link [here](https://codesandbox.io/s/pedantic-thunder-b6siqu?file=/src/App.js)

See, this AND does not have the combinators header:
![image](https://user-images.githubusercontent.com/24373382/215538424-d98e9b53-015c-4902-b50e-746da7c31378.png)

This one, has it:
![image](https://user-images.githubusercontent.com/24373382/215539095-7a6d1f83-37f1-4fa4-bc29-d0982edf94cc.png)

![image](https://user-images.githubusercontent.com/24373382/215539219-b74fe6ef-59ef-4bc6-ab98-5fffdf256c7c.png)

**Reply by jakeboone02** [reply]
You need to pick a format, either 1) use the standard format where `combinator` is at the group level and `independentCombinators` is `false`/`undefined`, or 2) use the "IC" format where a combinator name is placed between each rule/group in the `rules` array and `independentCombinators` is `true`. You're mixing the two.

### Option 1 (standard)

JSX:

```tsx
const App = () => (
  <QueryBuilder fields={fields} defaultQuery={initialQuery} />
);
```

Query format:

```json
{
  "combinator": "and",
  "rules": [
    { "field": "f1", "operator": "=", "value": "v1" },
    { "field": "f2", "operator": "=", "value": "v2" }
  ]
}
```

### Option 2 (IC)

JSX:

```tsx
const App = () => (
  <QueryBuilder independentCombinators fields={fields} defaultQuery={initialQuery} />
);
```

Query format:

```json
{
  "rules": [
    { "field": "f1", "operator": "=", "value": "v1" },
    "and",
    { "field": "f2", "operator": "=", "value": "v2" }
  ]
}
```

**Reply by keerthankrishna** [reply]
Thank you @jakeboone02 - I had a different understanding.

**Reply by keerthankrishna** [reply]
Is there an example you have using ` idgenerator ` prop ? I'm trying to generate a custom ID where I can get something like _"r-1", "r-2", "g-1", g-2"..._ instead of the ` math.random() ` output that comes out by default.

**Reply by jakeboone02** [reply]
`generateID` doesn't use `Math.random()` anymore unless it can't find a `crypto` module.  If you want to base the `id`s on the properties of each object, like whether it's a rule or a group, then `idGenerator` prop is not what you want. That will be passed no parameters, so it won't know what it's generating an `id` for.

I think what you probably want to do is use something like [`transformQuery`](https://react-querybuilder.js.org/docs/api/misc#transformquery) so you have more granular control. To generate `id`s in the format of your examples:

```ts
let ruleCount = 0;
let ruleGroupCount = 0;

const tq = transformQuery(initialQuery, {
  ruleProcessor: r => ({ ...r, id: `r-${++ruleCount}` }),
  ruleGroupProcessor: rg => ({ ...rg, id: `g-${++ruleGroupCount}` }),
});
```

**Reply by keerthankrishna** [reply]
Ah, okay - this works perfect.. thank you for this. This is what I was looking for!

**Reply by keerthankrishna** [reply]
Does the ` combinatorMap ` work the same way where I can send the selected data and have it be displayed within a custom export JSON ?
If I wanted the rule ID, the group ID, the fields, the operators and the combinators in a a custom array, can I add this ` combinatorMap ` specifically within the ` transformQuery ` and pass it to a header called "_combinator_" to fetch the selected data ? _(hope the question makes sense)_

I was able to already retrieve the fields and operators and with your previous comment was also able to generate an incremental rule/group ID.. and so was looking to add this combinator value too, hence the question.

**Reply by jakeboone02** [reply]
The question doesn't make sense to me, but because of that I'm pretty sure the answer is no.

**Reply by keerthankrishna** [reply]
Let me explain this further so that it's more clear and maybe you may have some ideas on the best approach or if this is feasible.
When I create a query, I'm trying to have ` QueryBuilder ` provide me JSON with all the query data which I can use to segregate and store in the table.

I'm looking at a JSON export after the query is created, which looks _somewhat_ like this :
```
[
  {
    rule_id:"r-1"
    group_id:"g-1"
    field_id:1
    field:"FirstName"
    operator_id:2
    operator:"="
    input_value:"Stev"
    rule-combinator:"and" //this is and because there is an and combinator between this rule and the next rule 
    group-combinator:"or" //this is or because there is an or combinator between the two groups
  },
 {
    rule_id:"r-2"
    group_id:"g-1"
    field_id:2
    field:"LastName"
    operator_id:2
    operator:"="
    value:"Smith"
    rule-combinator:"" //empty because there is no rule after this in this group
    group-combinator:"or"
  },
 {
    rule_id:"r-3"
    group_id:"g-2"
    field_id:1
    field:"FirstName"
    operator_id:2
    operator:"="
    value:"Markus"
    rule-combinator:"" //empty because there is no rule after this in this group
    group-combinator:"" //empty because there is no group after this
  },
]
```
Here's what the rule would look like, UI side, in ` QueryBuilder ` 
![image](https://user-images.githubusercontent.com/24373382/216522642-dce1b422-ac2f-44e6-b18a-f7c895cc1b56.png)

Let me know if you want to see this as an example to understand it better.

**Reply by jakeboone02** [reply]
Are you wanting to translate the standard query format into the format of your example JSON?

**Reply by keerthankrishna** [reply]
Yes, @jakeboone02 that is right. I want this because I am wanting to store this data into a table on the backend, so converting that into the JSON will help me store it easily.

**Reply by jakeboone02** [reply]
Normally I just recommend using `formatQuery(query, 'json_without_ids')` to serialize queries for storage. Why would you need to convert them into that new format?

**Reply by keerthankrishna** [reply]
I'm trying to store the data formed via the ` QueryBuilder ` rules into a SQL table so that I can use that to for an Update operation on the backend. That JSON will allow me to break the rule completely and store within a table, hence the ask - The only 2 items I am stuck right now are the _group combinators_ and the _rule combinators_ which I'm not able to get. All the other fields are retrievable.

Additionally, I use the ` QueryBuilder ` with ` independentCombinators ` set as _true_

**Reply by jakeboone02** [reply]
This library doesn't provide a tool to convert the query object into (or from) that format. Seems like you'll have to write a converter manually.

**Reply by keerthankrishna** [reply]
okay @jakeboone02 - I'll try to get this done manually.

**Comment by keerthankrishna** [comment]
@jakeboone02 - When using a Custom ` ValueEditor `, Is it possible to access all the rules from the ` ValueEditorProps ` ? I noticed each time a rule is created, the ` ValueEditorProps ` stores just that last rule but not the whole thing. Hope the question makes sense.

**Reply by jakeboone02** [reply]
What I usually recommend is to pass the query down through the `context` prop which is passed to every subcomponent. Something like this:

```tsx
<QueryBuilder
  query={query}
  onQueryChange={q => setQuery(q)}
  controlElements={{ valueEditor: CustomValueEditor }}
  context={{ query }}
/>
```

Then in your custom value editor:

```tsx
export const CustomValueEditor = (props: ValueEditorProps) => {
  const siblingRules = findPath(getParentPath(props.path), props.context.query).rules;

  // do something with siblingRules here...

  return <input {...whateverHere} />;
}
```

**Reply by keerthankrishna** [reply]
This works great.. Thank you @jakeboone02 for the direction!

**Comment by keerthankrishna** [comment]
@jakeboone02 Similar to how you pass data into the Custom ` ValueEditor `, is it possible to send it back out to the main component where the ` QueryBuilder ` component is called ?
**Example:**

- I have a parent component called "Form.js" where I am calling the main ` QueryBuilder ` component
- In the ` QueryBuilder ` component, I am calling a custom ` ValueEditor ` for which I have a seperate file called `customvalueeditor.js`
- Now, within `customvalueeditor.js`, I can get all the selection values
- If I want this selection values to be sent outside of the custom `valueEditor`, how do I do it ?
- Say for example here, _I want to know if "contains" has been selected and if yes, I want to know the values put into that rule_ **and** I want to access this in the parent component "Form.js"

**Reply by jakeboone02** [reply]
The question is this: _which_ value do you want to access? Your custom value editor will be rendered as many times are there are rules in your query, so the best way to check the `value` (or `operator`, which it seems like you're talking about since you mention "contains") is to either 1) find the rule with a particular `path` using `findPath`, or 2) recursively step through the entire query (which you already have access to in your parent component via the `onQueryChange` callback) and get _each_ value.

**Reply by keerthankrishna** [reply]
@jakeboone02  Thank you.. I think I can see the direction I need to take now. What would be the "_path_" in the parent component ?
When inside the custom ` ValueEditor ` component, my path is the ` ValueEditorProps.path ` but I'm not sure what would be the _path_ when I'm at the parent component.

**Reply by jakeboone02** [reply]
We should be more clear about this in the docs, actually. `path` is an array of integers specific to each rule and group within the query. It has no meaning in the context of the query as a whole (well, the root group has a `path` of `[]`) or in the parent component of the `<QueryBuilder />`.

Consider the following query. I've put the respective `path` as a comment above each object.

```jsonc
// []
{
  "combinator": "and",
  "rules": [
    // [0]
    { "field": "f1", "operator": "=", "value": "v1" },
    // [1]
    {
      "combinator": "or",
      "rules": [
        // [1, 0]
        { "field": "f2", "operator": "=", "value": "v2" },
        // [1, 1]
        { "field": "f3", "operator": "=", "value": "v3" }
      ]
    }
  ]
}
```

The `path` of the top-most rule in the query is `[0]`, `0` being the index of that rule within the root group's `rules` array. The `path` of the first sub-group in the query is `[1]`, again according to its index within the `rules` array. Each child rule of _that_ group has a path of `1` (the same index of its parent), followed by its own index.

When you run `findPath`, you pass 1) an array of numbers representing the `path` of the rule or group you want it to return, and 2) a query (i.e., a root group) to search. For example, given the query above:

```ts
findPath([1, 0], query)
```

would return this object:

```ts
{ field: "f2", operator: "=", value: "v2" }
```

Again, you need to identify _which_ rule value you want to access.

**Reply by keerthankrishna** [reply]
Thank you @jakeboone02 ! This makes it clear..
